pkg/ingester/user_tsdb.go

```diff
@@ -153,54 +164,100 @@ func (u *userTSDB) StartTime() (int64, error) {
+// having huge blocks.
+//
+// The input forcedMaxTime allows to specify the maximum timestamp of samples compacted from the
+// in-order Head. You can pass math.MaxInt64 to compact the entire in-order Head.
+func (u *userTSDB) compactHead(blockDuration, forcedCompactionMaxTime int64) error {
+	if ok, s := u.changeStateToForcedCompaction(active, forcedCompactionMaxTime); !ok {
 		return fmt.Errorf("TSDB head cannot be compacted because it is not in active state (possibly being closed or blocks shipping in progress): %s", s.String())
 	}
 
-	defer u.casState(forceCompacting, active)
+	defer u.changeState(forceCompacting, active)
 
-	// Ingestion of samples in parallel with forced compaction can lead to overlapping blocks,
-	// and possible invalidation of the references returned from Appender.GetRef().
-	// So we wait for existing in-flight requests to finish. Future push requests would fail until compaction is over.
-	u.pushesInFlight.Wait()
+	// Ingestion of samples with a time range overlapping with forced compaction can lead to overlapping blocks.
+	// For this reason, we wait for existing in-flight requests to finish, except the ones that have been intentionally
+	// allowed while forced compaction was in progress because they append samples newer than forcedMaxTime
+	// (requests appending samples older than forcedMaxTime will fail until forced compaction is completed).
+	u.inFlightAppendsWithoutForcedCompaction.Wait()
 
+	// Compact the TSDB head.
 	h := u.Head()
+	for {
+		blockMinTime, blockMaxTime, isValid, isLast := nextForcedHeadCompactionRange(blockDuration, h.MinTime(), h.MaxTime(), forcedCompactionMaxTime)
+		if !isValid {
+			break
+		}
 
-	minTime, maxTime := h.MinTime(), h.MaxTime()
-
-	for (minTime/blockDuration)*blockDuration != (maxTime/blockDuration)*blockDuration {
-		// Data in Head spans across multiple block ranges, so we break it into blocks here.
-		// Block max time is exclusive, so we do a -1 here.
-		blockMaxTime := ((minTime/blockDuration)+1)*blockDuration - 1
-		if err := u.db.CompactHead(tsdb.NewRangeHead(h, minTime, blockMaxTime)); err != nil {
+		if err := u.db.CompactHead(tsdb.NewRangeHead(h, blockMinTime, blockMaxTime)); err != nil {
 			return err
 		}
 
-		// Get current min/max times after compaction.
-		minTime, maxTime = h.MinTime(), h.MaxTime()
-	}
-	err := u.db.CompactHead(tsdb.NewRangeHead(h, minTime, maxTime))
-	if err != nil {
-		return err
+		// Do not check again if it was the last range.
+		if isLast {
+			break
+		}
 	}
+
 	return u.db.CompactOOOHead()
 }
 
+// nextForcedHeadCompactionRange computes the next TSDB head range to compact when a forced compaction
+// is triggered. If the returned isValid is false, then the returned range should not be compacted.
+func nextForcedHeadCompactionRange(blockDuration, headMinTime, headMaxTime, forcedMaxTime int64) (minTime, maxTime int64, isValid, isLast bool) {
```

Comment:
nit: `isLast bool` looks like unnecessary optimization to save one call to the function, and it can be wrong (if forcedMaxTime is aligned on block boundary). I'd drop it.

Reply:
It's not an optimization. The problem is that when compacting a partial head (so for the case of this PR) without that flag we may end up in a infinite loop.

Reply:
Since no pushes are allowed that would modify maxTime, why would we end up in an infinite loop? Are you talking about the case when minTime == maxTime?